native/cocos/platform/android/AndroidPlatform.cpp

```diff
@@ -25,79 +25,471 @@
+
+                if (actionMasked == AMOTION_EVENT_ACTION_DOWN ||
+                    actionMasked == AMOTION_EVENT_ACTION_POINTER_DOWN) {
+                    touchEvent.type = cc::TouchEvent::Type::BEGAN;
+                } else if (actionMasked == AMOTION_EVENT_ACTION_UP ||
+                           actionMasked == AMOTION_EVENT_ACTION_POINTER_UP) {
+                    touchEvent.type = cc::TouchEvent::Type::ENDED;
+                } else if (actionMasked == AMOTION_EVENT_ACTION_CANCEL) {
+                    touchEvent.type = cc::TouchEvent::Type::CANCELLED;
+                } else if (actionMasked == AMOTION_EVENT_ACTION_MOVE) {
+                    touchEvent.type = cc::TouchEvent::Type::MOVED;
+                } else {
+                    return false;
+                }
+
+                for (int i = 0; i < motionEvent->pointerCount; i++) {
+                    int id = motionEvent->pointers[i].id;
+                    float x = GameActivityPointerAxes_getX(&motionEvent->pointers[i]);
+                    float y = GameActivityPointerAxes_getY(&motionEvent->pointers[i]);
+                    touchEvent.touches.emplace_back(x, y, id);
+                }
+                _androidPlatform->dispatchEvent(touchEvent);
+                touchEvent.touches.clear();
+                return true;
+            }
+            return false;
+        }
+
+        bool cookGameActivityKeyEvent(GameActivityKeyEvent *keyEvent) {
+            if (keyEvent->keyCode == AKEYCODE_BACK && 0 == keyEvent->action) {
+                // back key was pressed
+                keyboardEvent.action = cc::KeyboardEvent::Action::PRESS;
+                keyboardEvent.key = static_cast<int>(KeyCode::MOBILE_BACK);
+                _androidPlatform->dispatchEvent(keyboardEvent);
+                return true;
+            }
+            return false;
+        }
+
+        void reportKeyState(int keyCode, bool state) {
+            bool wentDown = !keyState[keyCode] && state;
+            bool wentUp = keyState[keyCode] && !state;
+            keyState[keyCode] = state;
+
+            if (wentUp) {
+                keyboardEvent.key = keyCode;
+                keyboardEvent.action = cc::KeyboardEvent::Action::RELEASE;
+                _androidPlatform->dispatchEvent(keyboardEvent);
+            } else if (wentDown) {
+                keyboardEvent.key = keyCode;
+                keyboardEvent.action = cc::KeyboardEvent::Action::PRESS;
+                _androidPlatform->dispatchEvent(keyboardEvent);
+            }
+        }
+
+        int32_t getActiveGameControllerIndex() const {
+            return _gameControllerIndex;
         }
-        //std::this_thread::sleep_for(std::chrono::nanoseconds(100));
+
+        void setActiveGameControllerIndex(const int32_t controllerIndex) {
+            _gameControllerIndex = controllerIndex;
+        }
+
+        void handleAppCommand(int32_t cmd) {
+            switch (cmd) {
+                case APP_CMD_SAVE_STATE:
```

Comment:
Where are these commands issued? I don't seem to see
One more question, does HarmonyOS also support GameActivity?
I remember HarmonyOS was very similar to Android before

Reply:
APP_CMD_SAVE_STATE is the app lifecycle`s cmd, we just leave this code here.
HarmonyOS doesn't support GameActivity, It's package name doesn't support.

Reply:
Got it